### ISSUE TITLE
Live2D - Dynamic zoom or position is impossible 

### ISSUE THREAD

**Issue by PasixTheValkyrie** [renpy/renpy#5530]
Right now it's difficult to offer a dynamic experience with Live2D models.

I noted an upcoming change for the default_fade, but I figure such a change might be beneficial for most if not all of the properties of a Live2D model. As of right now, it's not possible to add a dynamic zoom or position as you'd expect for a normal sprite. This could be done in the Live2D model itself, but it's probably less desirable than being able to adjust it on a dynamic value.

Values such as default_fade, zoom and xpos and ypos are main contenders. The update_function can luckily be set to a function which you can override as you please, so that's something that'd be nice.

The pull for the default_fade I referenced
https://github.com/renpy/renpy/pull/5522

**Comment by renpytom** [comment]
I don't understand what you're asking for here. It should be possible to wrap a Live2D displayable in a Transform, and adjust things that way if you want to. The versions on the Live2D object are there for convenience.

**Comment by PasixTheValkyrie** [comment]
Well, that's exactly it. It's more convenient and cleaner if the option is there anyway. Wrapping it is a solution if you want to adjust it only once, but for dynamic zooming it is still inconvenient and still necessitates a loophole to get it done somewhat cleanly as far as I can tell.

I'll have to note I'm not super experienced with Ren'Py, but the documentation on its Live2D implementation are rather scarce and does not suggest modularity per model at all as I've also noticed for #5522. Working on a VN with multiple Live2D models at once can quickly get confusing.

**Comment by PasixTheValkyrie** [comment]
I might have somewhat of an example, although it's probably not exactly an answer to your question.

Right now, I have two subsets of animations, ones that loop, and ones that do not. Combined with the default_fade, this can cause issues where the sprite uses a non-looping animation, where it somehow tries to reset to an idle pose, while not needing to. In some cases this can be somewhat ignored if there's no major movement, like looking away and back. If you've got a sprite that moves its arms behind their back, for example, it's just a glitchy mess.

The only solution right now for that is to make a second model instance, which wouldn't be a massive problem, if Ren'Py didn't also lock whatever it does with the model to an existing one. Changes made to the second instance are ignored in favour of the first instance.

Not that massive of a problem either, just create a second folder with the exact same model, then instance that. We'll make it work, but convenient is anything but it so far.

Now we have a method of approach where you can do the following

Instance A Idle (looping)
Instance A Expression

Instance A Shrug (non-looping)
Pause for the duration of the non-looping animation
Switch out Instance A for Instance B (Since the default_fade makes the phase glitch)

Instance B Idle (looping)
Instance B Expression

It's really rather dirty code and that's not even including the capacity to do any zooms or repositioning dynamically.

I think the most straightforward solution to get rid of some of these steps is by allow different instances of the same model and making properties alterable.

A larger solution would effectively be animation sets, not too dissimilar from Unity's animation blending. Call on an animation, then once that animation is finished, return to idle.

I know it's wishful thinking though, I can not provide examples of how to implement this at all, but I hope my tangent somewhat explains what some of my struggles with the current Live2D implementation are.

**Comment by colinharvie** [comment]
You can create additional instances of your model by creating duplicate of your model3.json (but keeping the data the same, just have a secondary renamed file).

That said, I'm pretty confused as to what problem you are running into? Live2D Sprites should work exactly the same as any other displayable. You can zoom, move them around the scene, and even in 3d space, simply by giving them a transform as you would any other sprite?

**Comment by PasixTheValkyrie** [comment]
Frankly I do not know how to elaborate better than I already have. 

I spent yesterday fiddling around with it more but I cannot find a way to do both.

**Comment by Gouvernathor** [comment]
I don't understand what options are open in case of a normal sprite but closed in case of a live2d.
Why can't you just `show thelive2d at Transform(xpos=.7)` at some relevant part of your game - or even during an expression instead of that .7 ?

**Comment by PasixTheValkyrie** [comment]
Oh, I think I see the misunderstanding. The problem is not the transform at all. That's fine, but what I probably erronously referred to as dynamic is combining this with motions, triply so if these are either looping or non looping.

As mentioned in the longer comment, animation sets such as that are possible in Unity are simply not possible as far as I can find, and you cannot combine either with a transform as well.

Unless there's some third wrapping method that does all of it together, I cannot find a method that allows animation blending whilst subsequently I cannot find a way to circumvent this dynamically. The solution I have only works for static situations.

**Comment by renpytom** [comment]
Can you explain what you want without reference to other tools?

**Comment by PasixTheValkyrie** [comment]
I'll try again;

The keyword is animation *sets*. Basically when an animation is played and we want or need it to be followed up by something else, (a non-looping animation + a looping animation). Through this, we can essentially do something similar as to music channels where we can have multiple motions at the same time, examples for these are lipsync motions, including a non-looping shrug motion, ending in an idle motion.

Right now this isn't possible unless you throw down several lines and keep switching between live2d models within mere seconds, which is not only entirely unfriendly but also just unmanageable for a larger project.

This is inherently a Live2D feature, and not a Unity feature, but I keep referencing Unity as it has implemented Live2Ds features as such in its animation blending, which is a more visual example than I can give you.

For specific parameters, we can use blend_parameters through the update_function, which is absolutely powerful, but you can't force motions through it.

The reason I mentioned dynamic zoom and/or position and which most likely lead to the confusion is because if you try to loophole anything by forcing large series of changes in multiple models, you just simply cannot make it work with a transform. The model'd just glitch across the screen as it keeps getting reinstanced and it's just not a viable approach.

In retrospect however this is not the root cause nor the root concern, and I apologise for that confusion.

**Comment by colinharvie** [comment]
You can manually implement this with the update_function by grabbing the motion objects:
`theLive3Dcommon.motions.get("motionName")` and getting the calculated value for the time and and then blending the value in. like so:
```
motion_data = motion.get(t,0,False,False)
    for k, v in motion_data.items():
    kind, key = k
    factor, value = v
    if kind == "Parameter":
        d.blend_parameter(key, "Overwrite", value, factor)
```

However it is a bit of work to implement it. I do think more features for Live2D control would be beneficial for RenPy.
I've implemented a ton of additional things for Live2D in my game (though it's definitely a bit clunky), including a custom statement that does what I think you are describing:

`play motion spriteName [motion1,motion2] speed 1.0 factor 0.5 fadein 1.0 fadeout 1.0 loop sync continue` 

Is that the sort of thing you're looking for?

**Comment by PasixTheValkyrie** [comment]
Yes! I'm really glad you actually have code to support my ramblings as well.

I'm seriously interested in getting features such as these implemented for easier use.

Would you be willing to share some of the features you've been working on? I think it could help a lot already.

**Comment by colinharvie** [comment]
I can, though my code is real slipshod and at the moment interwoven with a bunch of other features, so it's not super easy to break out and reuse at the moment (though I'm planning to clean it up and make it easier to use, assuming none of the features get added to Ren'py in the mean time). I can share some but I don't think this issue thread is the best place for it. 

If you want though we can exchange emails, or discord (if you have one), and I can share some of the code snippets / we can hash out a good list of live2d features to ask the ren'py devs to implement with explainations.

**Comment by PasixTheValkyrie** [comment]
That'd be stellar, really. I'm simply Pasix on Discord, please feel free to add me there!

I think Ren'Py could definitely benefit from a stronger Live2D implementation so I'm all for hashing out such a list.

**Comment by Drachenfels** [comment]
@colinharvie Side note, how one implements custom statements in renpy? Do you have any tips or pointers on how to approach the problem? (My discord handle is #djdrachenfels if you don't mind sharing).

**Comment by colinharvie** [comment]
@Drachenfels Sure! I sent you a friend request on Discord.
